### ISSUE TITLE
[Reimplementation] Unable to reproduce results -- Training loss curves are similar

### ISSUE THREAD

**Issue by maximegmd** [neelsjain/NEFTune#1]
Hello,

We implemented hijacking for mistral and llama in Axolotl (see PR: https://github.com/OpenAccess-AI-Collective/axolotl/pull/721 ) but are unable to reproduce your results, our loss curves pretty much match the version without noisy embeddings. We then copied the implementation in your code instead of using the one in the readme and same results.

We have tried:
- With/Without sample packing
- Shorter context (512 like in your paper)
- Llama 2 and Mistral 7b

Could you help us understand what we are doing wrong?

**Comment by neelsjain** [comment]
Could you give more details? (i.e which dataset, model, hyperparameters, etc)

There should slight loss differences but nothing major from what we saw. Can you share these the loss numbers?

FYI -- traveling the next couple days so response timing might be delayed

**Comment by winglian** [comment]
@neelsjain I used a GPT-4 version of the alpaca dataset (https://huggingface.co/datasets/vicgalle/alpaca-gpt4), and was testing with a ShearedLlama 1.7b. In addition, we tried [a second implementation](https://github.com/OpenAccess-AI-Collective/axolotl/compare/neft-v2?expand=1) based on [the published training code](https://github.com/neelsjain/NEFTune/blob/main/experiment_code/train.py#L148-L159) rather than the published algorithm and we had the same results.

**Comment by neelsjain** [comment]
What were the AlpacaEval scores? Can you share the beginning of training loss values? We did not try this on models smaller than 7b.

**Comment by winglian** [comment]
we haven't done any alpaca evals yet. When we say same results, we don't see a vary large deviation in the train loss, like the chart you have in the paper, but rather the train losses overlap very closely to each other. 

I did some more experimentation and noticed that this is likely due to the fact that I generally already use weight decay when I finetune, and NEFT doesn't provide much additional regularization over weight decay.

**Comment by neelsjain** [comment]
I see. Our loss curves are very similar as well (almost the same). OpenLLM Leaderboard tasks are also very similar in scores. The place we where see the difference is in the generations of the models. I think you guys should at least look at a couple generations. The folks at HF seemed to have been able to train Mistral-7B on the guanaco dataset and saw a difference of ~25% on MT-Bench.

We did add weight decay of 0.1 for our 70B model run (copying the hparm from the LLaMA-2 paper) and noticed improvement when training on Evol-Instruct-70k. We have a table in the paper.

**Comment by sadransh** [comment]
@winglian, I see the same thing with different alpha parameters; both train/eval loss curves are almost the same!

**Comment by neelsjain** [comment]
@sadransh, are the generations from the model identical? I might just make clear in the README.md that this is the expected behavior (i.e. loss curves look similar during training). The training loss should be higher when the noise is turned off but they are similar when noise is turned on. The figures in the analysis section of the paper show the losses for the model trained on the Alpaca dataset when noise is turned off. 

It is worth mentioning that the noise level is quite mild as you are dividing by the square root of the embedding dimension and sequence length.

**Comment by sadransh** [comment]
@neelsjain Thanks for the nice work! 

I am doing experiments, so far what I have seen is (2-3% improve and 2-3% drop on similar checkpoint on test set) (alpha is 5)

train loss is looks like this, and become more similar on higher steps. 
<img width="457" alt="image" src="https://github.com/neelsjain/NEFTune/assets/12527824/4decd4d0-7666-44c5-be03-31d554d94f3d"> 
red is alpha 15, gray no alpha, and pink alpha is 5 

I followed the implementation of trl HF, and not 100% percent sure if it is correct. 

eval loss looks like: 
<img width="393" alt="image" src="https://github.com/neelsjain/NEFTune/assets/12527824/7e5cc02b-7609-4c43-a7a7-7bb7bdd40844">

Any thoughts? 

I need to look at generations manually!

**Comment by neelsjain** [comment]
Thanks for the kind words. 

Yes, this looks about right! I think we should wait and look the generations. The HF implementation looked fine to me at first glance. The folks at HF found that works better for weaker datasets, which to me is not at all surprising. IMO there is an upper limit how good the "style" of the generations are, which is what AlpacaEval mainly looks for. To be honest, I think NEFTune can be improved and probably has faults that we have not found yet. However, the idea of augmentation and thinking optimization process of the SFT for instruction tuning is an exciting direction of research.

**Comment by sadransh** [comment]
One more question, @neelsjain 
What are your stopping criteria and checkpoint selection logic for these experiments? Are you stopping based on minimum evaluation loss? Or is there a different logic?

**Comment by neelsjain** [comment]
We did 3 epochs. All the details are in the paper in the Appendix.
